Property definition that impacts generated AssemblyInfo doesn't support CData

_From @NTaylorMullen on November 1, 2016 18:30_

Adding a `<Description>` to my csproj that looks like:
```XML
<Description><![CDATA[Line 1
Line 2
Line 3]]></Description>
```

Results in the following error when trying to build:

```
MSBUILD : error MSB4025: The project file could not be loaded. Name cannot begin with the '<' character, hexadecimal value 0x3C. Line 7, position 44.
```

_Copied from original issue: dotnet/sdk#347_

Project File (from `dotnet new`, and then dded the description):

```XML
<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" />
  
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp1.0</TargetFramework>
    <Description><![CDATA[Line 1
Line 2
Line 3]]></Description>
  </PropertyGroup>

  <ItemGroup>
    <Compile Include="**\*.cs" />
    <EmbeddedResource Include="**\*.resx" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.NETCore.App">
      <Version>1.0.1</Version>
    </PackageReference>
    <PackageReference Include="Microsoft.NET.Sdk">
      <Version>1.0.0-alpha-20161104-2</Version>
      <PrivateAssets>All</PrivateAssets>
    </PackageReference>
  </ItemGroup>
  
  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
</Project>
```

Then trying to run results in this:
```
C:\Users\nimullen\Documents\temp\msbuild> dotnet restore
  Restoring packages for C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj...
  Writing lock file to disk. Path: C:\Users\nimullen\Documents\temp\msbuild\obj\project.assets.json
  Generating MSBuild file C:\Users\nimullen\Documents\temp\msbuild\obj\msbuild.csproj.nuget.g.targets.
  Generating MSBuild file C:\Users\nimullen\Documents\temp\msbuild\obj\msbuild.csproj.nuget.g.props.
  Restore completed in 2436.5539ms for C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj.

  NuGet Config files used:
      C:\Users\nimullen\AppData\Roaming\NuGet\NuGet.Config

  Feeds used:
      https://www.nuget.org/api/v2/
C:\Users\nimullen\Documents\temp\msbuild> dotnet run
obj\Debug\netcoreapp1.0\msbuild.AssemblyInfo.cs(8,59): error CS1010: Newline in constant [C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj]
obj\Debug\netcoreapp1.0\msbuild.AssemblyInfo.cs(8,66): error CS1003: Syntax error, ',' expected [C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj]
obj\Debug\netcoreapp1.0\msbuild.AssemblyInfo.cs(9,6): error CS1003: Syntax error, ',' expected [C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj]
obj\Debug\netcoreapp1.0\msbuild.AssemblyInfo.cs(9,7): error CS1003: Syntax error, ',' expected [C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj]
obj\Debug\netcoreapp1.0\msbuild.AssemblyInfo.cs(10,6): error CS1003: Syntax error, ',' expected [C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj]
obj\Debug\netcoreapp1.0\msbuild.AssemblyInfo.cs(10,7): error CS1003: Syntax error, ',' expected [C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj]
obj\Debug\netcoreapp1.0\msbuild.AssemblyInfo.cs(10,7): error CS1010: Newline in constant [C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj]
obj\Debug\netcoreapp1.0\msbuild.AssemblyInfo.cs(16,69): error CS1026: ) expected [C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj]
obj\Debug\netcoreapp1.0\msbuild.AssemblyInfo.cs(16,69): error CS1003: Syntax error, ']' expected [C:\Users\nimullen\Documents\temp\msbuild\msbuild.csproj]

The build failed. Please fix the build errors and run again.
```

On closer inspection it looks like the system is generating the msbuild assembly info file incorrectly. It looks like:
```C#
// Generated by the MSBuild WriteCodeFragment class.

using System;
using System.Reflection;

[assembly: System.Reflection.AssemblyCompanyAttribute("msbuild")]
[assembly: System.Reflection.AssemblyConfigurationAttribute("Debug")]
[assembly: System.Reflection.AssemblyDescriptionAttribute("Line 1
Line 2
Line 3")]
[assembly: System.Reflection.AssemblyFileVersionAttribute("1.0.0.0")]
[assembly: System.Reflection.AssemblyInformationalVersionAttribute("1.0.0")]
[assembly: System.Reflection.AssemblyProductAttribute("msbuild")]
[assembly: System.Reflection.AssemblyTitleAttribute("msbuild")]
[assembly: System.Reflection.AssemblyVersionAttribute("1.0.0.0")]
[assembly: System.Resources.NeutralResourcesLanguageAttribute("en")]
```

Lastly msbuild version is `15.1.0.0` and `dotnet --info`:
```
.NET Command Line Tools (1.0.0-preview4-004079)

Product Information:
 Version:            1.0.0-preview4-004079
 Commit SHA-1 hash:  43dfa6b8ba

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
```

@rainersigwald must have also tested a property that impacts the assembly info. I've updated the title.

--- a/src/XMakeTasks/UnitTests/WriteCodeFragment_Tests.cs
+++ b/src/XMakeTasks/UnitTests/WriteCodeFragment_Tests.cs
@@ -429,7 +429,7 @@ namespace Microsoft.Build.UnitTests
             File.Delete(task.OutputFile.ItemSpec);
         }
 
-        public static string EscapedLineSeparator => (NativeMethodsShared.IsWindows ? "\\r" : "") + "\\n";
+        public static string EscapedLineSeparator => NativeMethodsShared.IsWindows ? "\\r\\n" : "\\n";
 
         /// <summary>
         /// Multi line argument values should cause a verbatim string to be used

--- a/src/XMakeTasks/UnitTests/WriteCodeFragment_Tests.cs
+++ b/src/XMakeTasks/UnitTests/WriteCodeFragment_Tests.cs
@@ -429,6 +429,86 @@ namespace Microsoft.Build.UnitTests
             File.Delete(task.OutputFile.ItemSpec);
         }
 
+        public static string EscapedLineSeparator => (NativeMethodsShared.IsWindows ? "\\r" : "") + "\\n";
+
+        /// <summary>
+        /// Multi line argument values should cause a verbatim string to be used
+        /// </summary>
+        [Fact]
+        public void MultilineAttributeCSharp()
+        {
+            var lines = new[] { "line 1", "line 2", "line 3" };
+            var multilineString = String.Join(Environment.NewLine, lines);
+
+            WriteCodeFragment task = new WriteCodeFragment();
+            MockEngine engine = new MockEngine(true);
+            task.BuildEngine = engine;
+            TaskItem attribute = new TaskItem("System.Reflection.AssemblyDescriptionAttribute");
+            attribute.SetMetadata("_Parameter1", multilineString);
+            attribute.SetMetadata("Description", multilineString);
+            task.AssemblyAttributes = new TaskItem[] { attribute };
+            task.Language = "c#";
+            task.OutputDirectory = new TaskItem(Path.GetTempPath());
+            bool result = task.Execute();
+
+            Assert.Equal(true, result);
+
+            string content = File.ReadAllText(task.OutputFile.ItemSpec);
+            Console.WriteLine(content);
+
+            var csMultilineString = lines.Aggregate((l1, l2) => l1 + EscapedLineSeparator + l2);
+            CheckContentCSharp(content, $"[assembly: System.Reflection.AssemblyDescriptionAttribute(\"{csMultilineString}\", Description=\"{csMultilineString}\")]");
+
+            File.Delete(task.OutputFile.ItemSpec);
+        }
+
+        private static readonly string VBCarriageReturn = "Global.Microsoft.VisualBasic.ChrW(13)";
+        private static readonly string VBLineFeed = "Global.Microsoft.VisualBasic.ChrW(10)";
+        private static readonly string WindowsNewLine = $"{VBCarriageReturn}&{VBLineFeed}";
+
+        public static readonly string VBLineSeparator =
+#if FEATURE_CODEDOM
+            WindowsNewLine;
+#else
+            NativeMethodsShared.IsWindows
+                ? WindowsNewLine
+                : VBLineFeed;
+#endif
+
+        /// <summary>
+        /// Multi line argument values should cause a verbatim string to be used
+        /// </summary>
+        [Fact]
+        public void MultilineAttributeVB()
+        {
+            var lines = new []{ "line 1", "line 2", "line 3" };
+            var multilineString = String.Join(Environment.NewLine, lines);
+
+            WriteCodeFragment task = new WriteCodeFragment();
+            MockEngine engine = new MockEngine(true);
+            task.BuildEngine = engine;
+            TaskItem attribute = new TaskItem("System.Reflection.AssemblyDescriptionAttribute");
+            attribute.SetMetadata("_Parameter1", multilineString);
+            attribute.SetMetadata("Description", multilineString);
+            task.AssemblyAttributes = new TaskItem[] { attribute };
+            task.Language = "visualbasic";
+            task.OutputDirectory = new TaskItem(Path.GetTempPath());
+            bool result = task.Execute();
+
+            Assert.Equal(true, result);
+
+            string content = File.ReadAllText(task.OutputFile.ItemSpec);
+            Console.WriteLine(content);
+
+            var vbMultilineString = lines
+                .Select(l => $"\"{l}\"")
+                .Aggregate((l1, l2) => $"{l1}&{VBLineSeparator}&{l2}");
+
+            CheckContentVB(content, $"<Assembly: System.Reflection.AssemblyDescriptionAttribute({vbMultilineString}, Description:={vbMultilineString})>");
+
+            File.Delete(task.OutputFile.ItemSpec);
+        }
+
         /// <summary>
         /// Some attributes only allow positional constructor arguments.
         /// To set those, use metadata names like "_Parameter1", "_Parameter2" etc.

--- a/src/XMakeTasks/WriteCodeFragment.cs
+++ b/src/XMakeTasks/WriteCodeFragment.cs
@@ -303,7 +303,7 @@ namespace Microsoft.Build.Tasks
 
                     foreach (ITaskItem attributeItem in AssemblyAttributes)
                     {
-                        string args = GetAttributeArguments(attributeItem, "=");
+                        string args = GetAttributeArguments(attributeItem, "=", QuoteSnippetStringCSharp);
                         if (args == null) return null;
 
                         code.AppendLine(string.Format($"[assembly: {attributeItem.ItemSpec}({args})]"));
@@ -326,7 +326,7 @@ namespace Microsoft.Build.Tasks
 
                     foreach (ITaskItem attributeItem in AssemblyAttributes)
                     {
-                        string args = GetAttributeArguments(attributeItem, ":=");
+                        string args = GetAttributeArguments(attributeItem, ":=", QuoteSnippetStringVisualBasic);
                         if (args == null) return null;
 
                         code.AppendLine(string.Format($"<Assembly: {attributeItem.ItemSpec}({args})>"));
@@ -343,7 +343,7 @@ namespace Microsoft.Build.Tasks
             return haveGeneratedContent ? code.ToString() : string.Empty; 
         }
 
-        private string GetAttributeArguments(ITaskItem attributeItem, string namedArgumentString)
+        private string GetAttributeArguments(ITaskItem attributeItem, string namedArgumentString, Func<string, string> quoteString)
         {
             // Some attributes only allow positional constructor arguments, or the user may just prefer them.
             // To set those, use metadata names like "_Parameter1", "_Parameter2" etc.
@@ -357,7 +357,7 @@ namespace Microsoft.Build.Tasks
             foreach (DictionaryEntry entry in customMetadata)
             {
                 string name = (string) entry.Key;
-                string value = entry.Value is string ? $@"""{entry.Value}""" : entry.Value.ToString();
+                string value = entry.Value is string ? quoteString(entry.Value.ToString()) : entry.Value.ToString();
 
                 if (name.StartsWith("_Parameter", StringComparison.OrdinalIgnoreCase))
                 {
@@ -403,6 +403,215 @@ namespace Microsoft.Build.Tasks
             }
 
             return string.Join(", ", orderedParameters.Union(namedParameters).Where(p => !string.IsNullOrWhiteSpace(p)));
+        }
+
+        private const int MaxLineLength = 80;
+
+        // copied from Microsoft.CSharp.CSharpCodeProvider
+        private string QuoteSnippetStringCSharp(string value)
+        {
+            // If the string is short, use C style quoting (e.g "\r\n")
+            // Also do it if it is too long to fit in one line
+            // If the string contains '\0', verbatim style won't work.
+            if (value.Length < 256 || value.Length > 1500 || (value.IndexOf('\0') != -1))
+                return QuoteSnippetStringCStyle(value);
+
+            // Otherwise, use 'verbatim' style quoting (e.g. @"foo")
+            return QuoteSnippetStringVerbatimStyle(value);
+        }
+
+        // copied from Microsoft.CSharp.CSharpCodeProvider
+        private string QuoteSnippetStringCStyle(string value)
+        {
+            StringBuilder b = new StringBuilder(value.Length + 5);
+
+            b.Append("\"");
+
+            int i = 0;
+            while (i < value.Length)
+            {
+                switch (value[i])
+                {
+                    case '\r':
+                        b.Append("\\r");
+                        break;
+                    case '\t':
+                        b.Append("\\t");
+                        break;
+                    case '\"':
+                        b.Append("\\\"");
+                        break;
+                    case '\'':
+                        b.Append("\\\'");
+                        break;
+                    case '\\':
+                        b.Append("\\\\");
+                        break;
+                    case '\0':
+                        b.Append("\\0");
+                        break;
+                    case '\n':
+                        b.Append("\\n");
+                        break;
+                    case '\u2028':
+                    case '\u2029':
+                        b.Append("\\n");
+                        break;
+
+                    default:
+                        b.Append(value[i]);
+                        break;
+                }
+
+                if (i > 0 && i%MaxLineLength == 0)
+                {
+                    //
+                    // If current character is a high surrogate and the following 
+                    // character is a low surrogate, don't break them. 
+                    // Otherwise when we write the string to a file, we might lose 
+                    // the characters.
+                    // 
+                    if (Char.IsHighSurrogate(value[i])
+                        && (i < value.Length - 1)
+                        && Char.IsLowSurrogate(value[i + 1]))
+                    {
+                        b.Append(value[++i]);
+                    }
+
+                    b.Append("\" +");
+                    b.Append(Environment.NewLine);
+                    b.Append('\"');
+                }
+                ++i;
+            }
+
+            b.Append("\"");
+
+            return b.ToString();
+        }
+
+        // copied from Microsoft.CSharp.CSharpCodeProvider
+        private string QuoteSnippetStringVerbatimStyle(string value)
+        {
+            StringBuilder b = new StringBuilder(value.Length + 5);
+
+            b.Append("@\"");
+
+            for (int i = 0; i < value.Length; i++)
+            {
+                if (value[i] == '\"')
+                    b.Append("\"\"");
+                else
+                    b.Append(value[i]);
+            }
+
+            b.Append("\"");
+
+            return b.ToString();
+        }
+
+        // copied from Microsoft.VisualBasic.VBCodeProvider
+        private string QuoteSnippetStringVisualBasic(string value)
+        {
+            StringBuilder b = new StringBuilder(value.Length + 5);
+
+            bool fInDoubleQuotes = true;
+
+            b.Append("\"");
+
+            int i = 0;
+            while (i < value.Length)
+            {
+                char ch = value[i];
+                switch (ch)
+                {
+                    case '\"':
+                    // These are the inward sloping quotes used by default in some cultures like CHS. 
+                    // VBC.EXE does a mapping ANSI that results in it treating these as syntactically equivalent to a
+                    // regular double quote.
+                    case '\u201C':
+                    case '\u201D':
+                    case '\uFF02':
+                        EnsureInDoubleQuotes(ref fInDoubleQuotes, b);
+                        b.Append(ch);
+                        b.Append(ch);
+                        break;
+                    case '\r':
+                        EnsureNotInDoubleQuotes(ref fInDoubleQuotes, b);
+                        if (i < value.Length - 1 && value[i + 1] == '\n')
+                        {
+                            b.Append("&Global.Microsoft.VisualBasic.ChrW(13)&Global.Microsoft.VisualBasic.ChrW(10)");
+                            i++;
+                        }
+                        else
+                        {
+                            b.Append("&Global.Microsoft.VisualBasic.ChrW(13)");
+                        }
+                        break;
+                    case '\t':
+                        EnsureNotInDoubleQuotes(ref fInDoubleQuotes, b);
+                        b.Append("&Global.Microsoft.VisualBasic.ChrW(9)");
+                        break;
+                    case '\0':
+                        EnsureNotInDoubleQuotes(ref fInDoubleQuotes, b);
+                        b.Append("&Global.Microsoft.VisualBasic.ChrW(0)");
+                        break;
+                    case '\n':
+                    case '\u2028':
+                    case '\u2029':
+                        EnsureNotInDoubleQuotes(ref fInDoubleQuotes, b);
+                        b.Append("&Global.Microsoft.VisualBasic.ChrW(10)");
+                        break;
+                    default:
+                        EnsureInDoubleQuotes(ref fInDoubleQuotes, b);
+                        b.Append(value[i]);
+                        break;
+                }
+
+                if (i > 0 && i%MaxLineLength == 0)
+                {
+                    //
+                    // If current character is a high surrogate and the following 
+                    // character is a low surrogate, don't break them. 
+                    // Otherwise when we write the string to a file, we might lose 
+                    // the characters.
+                    // 
+                    if (Char.IsHighSurrogate(value[i])
+                        && (i < value.Length - 1)
+                        && Char.IsLowSurrogate(value[i + 1]))
+                    {
+                        b.Append(value[++i]);
+                    }
+
+                    if (fInDoubleQuotes)
+                        b.Append("\"");
+                    fInDoubleQuotes = true;
+
+                    b.Append("& _ ");
+                    b.Append(Environment.NewLine);
+                    b.Append('\"');
+                }
+                ++i;
+            }
+
+            if (fInDoubleQuotes)
+                b.Append("\"");
+
+            return b.ToString();
+        }
+
+        private void EnsureNotInDoubleQuotes(ref bool fInDoubleQuotes, StringBuilder b)
+        {
+            if (!fInDoubleQuotes) return;
+            b.Append("\"");
+            fInDoubleQuotes = false;
+        }
+
+        private void EnsureInDoubleQuotes(ref bool fInDoubleQuotes, StringBuilder b)
+        {
+            if (fInDoubleQuotes) return;
+            b.Append("&\"");
+            fInDoubleQuotes = true;
         }
     }
 }

